### PR TITLE
[ADP-3344] Implement REST-like `IO` interface for Deposit Wallet

### PR DESCRIPTION
**Pull request by HeinrichApfelmus**
This pull requests implements a REST-like `IO` interface for the Deposit Wallet. This interface corresponds more closely to an HTTP interface.

Essentially, the main difference between `Cardano.Wallet.Deposit.IO` and `Cardano.Wallet.Deposit.REST` is that the latter postpones the need to initialize the wallet. Instead of working with a `WalletInstance`, the latter module now works with a `WalletResource`, which is roughly a `TVar (Maybe WalletInstance)`. This mutable variable starts with `Nothing` and the function `putWallet` can be used to initialize the wallet, putting a `Just` in the mutable variable.

In turn, a module `Cardano.Wallet.Deposit.IO.Resource` implements a data type `Resource` that encapsulate the idea of `TVar (Maybe WalletInstance)` and is (reasonably) safe to use in a concurrent environment.

### Comments

* Not implemented yet: `withWalletResource` inspects the database directory and loads a wallet if necessary.

### Issue Number

ADP-3344

### CODE REVIEW

File: lib/crypto-primitives/src/Cryptography/Hash/Blake.hs [hunk 1]
```diff
@@ -41,5 +42,8 @@ blake2b256 = BA.convert . hash @_ @Blake2b_256
 blake2b224 :: ByteArrayAccess a => a -> ByteString
 blake2b224 = BA.convert . hash @_ @Blake2b_224
 
+blake2b160 :: ByteArrayAccess a => a -> ByteString
+blake2b160 = BA.convert . hash @_ @Blake2b_160
```
Comment (paweljakubas) [inline lib/crypto-primitives/src/Cryptography/Hash/Blake.hs:46]: :+1:

File: lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/IO.hs [hunk 2]
```diff
@@ -108,44 +126,57 @@ readWalletState WalletInstance{walletState} =
...
     -> (WalletInstance -> IO a)
     -> IO a
-withWalletInit env@WalletEnv{..} xpub knownCustomerCount action = do
-    walletState <- DBVar.initDBVar database
-        $ Wallet.fromXPubAndGenesis xpub knownCustomerCount genesisData
-    withWalletDBVar env walletState action
+withWalletInit
+    env@WalletEnv
+        { bootEnv = WalletBootEnv{genesisData}
+        , ..
```
Comment (paweljakubas) [inline lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/IO.hs:140]: I was not aware of this feature :sweat_smile:

File: lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/IO/Resource.hs [hunk 3]
```diff
@@ -0,0 +1,165 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RankNTypes #-}
+{-# HLINT ignore "Use void" #-}
```
Comment (paweljakubas) [inline lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/IO/Resource.hs]: I would put that under LANG and OPTIONS in separate section

File: lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/IO/Resource.hs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+-- | Make a 'Resource' that can be initialized later.
+--
+-- Once the 'Resource' has been initialized,
+-- it will also be cleaned up once the 'withResource' function has finished.
+--
+-- If the 'Resource' vanishes because of an exception,
+-- the 'withResource' will /not/ be interrupted.
+-- You can use 'getStatus' to poll the current status.
+withResource
```
Comment (paweljakubas) [inline lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/IO/Resource.hs:89]: I wonder if we should not put `withResource` in the separate module and package - seems like very reusable thing. Does not `local cluster` use similar approach?

File: lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/REST.hs [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+    = ErrNoWallet Resource.ErrResourceMissing
+    | ErrYesWallet (Resource.ErrResourceExists WalletIO.WalletInstance)
+
+instance Show ErrWalletResource where
+    show = \case
+        ErrNoWallet e -> case e of
+            ErrNotInitialized -> "Wallet is not initialized"
+            ErrStillInitializing -> "Wallet is still initializing"
+            ErrVanished e' -> "Wallet absent and vanished: " <> show e'
+        ErrYesWallet e -> case e of
```
Comment (paweljakubas) [inline lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/REST.hs]:
```suggestion
        ErrWalletPresent e -> case e of
```

**Review by paweljakubas — APPROVED** [review]
Nice work!

File: lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/REST.hs [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+    ExceptT
+        $ first ErrNoWallet <$> Resource.onResource action resource
+
+{-----------------------------------------------------------------------------
+    Initialization
+------------------------------------------------------------------------------}
+
+-- | Prefix for deposit wallets on disk.
+depostiPrefix :: String
+depostiPrefix = "deposit-"
```
Comment (HeinrichApfelmus) [inline lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/REST.hs]:
```suggestion
depositPrefix :: String
depositPrefix = "deposit-"
```

(typo)

**Comment by HeinrichApfelmus** [comment]
Looks good to me as well, thank you! 👍

The error type `e` is probably not worth tracking in the type system, hence my preference for putting this stuff under the rug in `SomeException`. However, distinguishing between `Vanished` and `FailedToInitialize` is indeed useful information for a user who tries to start the wallet. In any case, having an `e` parameter is alright.

**Pull request merged**